### PR TITLE
added BYOS ("Bring Your Own SERVICE_ACCOUNT") for GCS authentication  

### PR DESCRIPTION
**Pull request by akashShanmugraj**
#### Description

This pull request allows the application to authenticate its Google Cloud Storage buckets through given SERVICE_ACCOUNT credential. The path to JSON file containing the service account credentials must be pre-set to the environment variable `GCP_CREDENTIAL_PATH`

#### Additional information...

- [] Have you run the test suite?
  Many changes need to be tested. Please run the test suite
  and include the output of ```cd test && ./problems```.

I ran `cargo test`.  All tests were successful.  `cargo build` was also successful. 
But I was not able to verify if it actually works

### CODE REVIEW

**Comment by akashShanmugraj** [comment]
The `./taskchampion-sync-server` executable spins up an Actix Webserver and I am not sure which routes I should visit to test it with my account

**Comment by djmitche** [comment]
> The `./taskchampion-sync-server` executable spins up an Actix Webserver and I am not sure which routes I should visit to test it with my account

Don't worry about that -- it's not relevant for this project.

File: taskchampion/lib/src/lib.rs [hunk 1]
```diff
@@ -129,7 +129,6 @@ mod util;
 pub mod annotation;
 pub use annotation::*;
 pub mod atomic;
-pub use atomic::*;
```
Comment (djmitche) [inline taskchampion/lib/src/lib.rs:133]: Can you make a separate PR for this?

Comment (akashShanmugraj) [reply taskchampion/lib/src/lib.rs:133]: Okay!

File: taskchampion/taskchampion/src/server/cloud/gcp.rs [hunk 2]
```diff
@@ -241,13 +243,21 @@ mod tests {
     /// Note that the Rust test runner will still show "ok" for the test, as there is no way to
     /// indicate anything else.
     fn make_service() -> Option<(GcpService, impl Fn(&str) -> Vec<u8>)> {
-        let Ok(bucket) = std::env::var("GCP_TEST_BUCKET") else {
-            return None;
+        let bucket = match std::env::var("GCP_TEST_BUCKET") {
+            Ok(val) => val,
+            Err(_) => return None,
+        };
```
Comment (djmitche) [inline taskchampion/taskchampion/src/server/cloud/gcp.rs]: Why did this change from the `let Ok(..) = .. else { .. }` syntax?

**Review by djmitche — CHANGES_REQUESTED** [review]
This looks good! I think it needs three more things:

1. Documentation update in `doc/man/task-sync.5.in`
2. Make the credential path optional. In Rust, this should be `Option<..>`, and that means the FFI needs to use TCStringType::Null to represent the `None` variant.
3. Add support for supplying this parameter from `taskrc`. If you look for `sync.gcp.bucket`, you should find all of the places that need to be updated.

File: taskchampion/taskchampion/src/server/cloud/gcp.rs (hunk 2 above)
Comment (akashShanmugraj) [reply taskchampion/taskchampion/src/server/cloud/gcp.rs]: Ah I thought this was a bit better in terms of exception handling, I'll change it back immediately

**Comment by akashShanmugraj** [comment]
> 2\. Make the credential path optional. In Rust, this should be `Option<..>`,

I tried to add the optional modifier to `credentialpath` but I am not sure how to fix this warning:

Screenshot 2024-01-21 at 6 01 00 AM

Plus I hope my string conversion here is correct for None too?
```rust
let credentialpath = match credentialpath {
                Some(path) => unsafe { TCString::val_from_arg(path) }.into_string()?,
                None => String::new(), // or provide a default path
            };
```

**Comment by djmitche** [comment]
You can't pass an `Optional<..>` value from C, so you'll need to keep `tc_server_new_gcp` taking a TCString argument. However, TCString has an `is_null` method which you can use to convert it to either `Some(String)` or `None`. Something like
```
let credentialpath = if credentialpath.is_null() {
  None
} else {
  // SAFETY: ..
  Some(unsafe { TCString::val_from_arg(path) }.into_string()?)
}
```

**Comment by akashShanmugraj** [comment]
Am I handling the `Option<String>` correctly?

```rust
let credentialpath = if credentialpath.is_null() {
                None
            } else {
                Some(unsafe { TCString::val_from_arg(credentialpath) }.into_string()?)
            };

            let credentialpath = credentialpath.unwrap();
```

**Comment by akashShanmugraj** [comment]
> 3\. If you look for `sync.gcp.bucket`, you should find all of the places that need to be updated.

Where should I find the `sync.gcp.bucket`? It is a file?

**Comment by djmitche** [comment]
> > 3. If you look for `sync.gcp.bucket`, you should find all of the places that need to be updated.
> 
> Where should I find the `sync.gcp.bucket`? It is a file?

It's an existing configuration parameter that is handled similarly to the new one.

Try
```
git grep sync.gcp.bucket
```
to find where it is referenced.

**Comment by djmitche** [comment]
> Am I handling the `Option<String>` correctly?

It's hard to say without seeing the patch. The compiler will probably tell you if you're not :)

> ```rust
>             let credentialpath = credentialpath.unwrap();
> ```

This doesn't seem right -- the statement before it carefully constructs an Option<String>, so it's odd to immediately unwrap that. It will panic in the `None` case!

**Comment by akashShanmugraj** [comment]
But if I dont unwrap it, I am getting mismatched type error

Screenshot 2024-01-21 at 7 21 36 AM

I am not sure how to handle it

**Comment by akashShanmugraj** [comment]
> It's an existing configuration parameter that is handled similarly to the new one.

Oh okay 
Now I should create a new parameter say `sync.gcp.credentials` at all the places where existing parameter `sync.gcp.bucket` is already defined, add and define that configuration everywhere.

**Comment by akashShanmugraj** [comment]
Maybe something like this then?
```rust
            let credpath:String;

            if credentialpath.is_null() {
                credpath = String::from("");
            } else {
                credpath = unsafe { TCString::val_from_arg(credentialpath) }.into_string()?;
            }
```

File: src/Context.cpp [hunk 3]
```diff
@@ -283,6 +283,7 @@ std::string configurationDefaults =
   "#sync.server.encryption_secret                 # Encryption secret for sync to a server\n"
   "#sync.server.origin                            # Origin of the sync server\n"
   "#sync.local.server_dir                         # Directory for local sync\n"
+  "#sync.gcp.credentialpath                       # Path to JSON file containing credentials to authenticate GCP Sync\n"
```
Comment (djmitche) [inline src/Context.cpp]: The configuration parameters have `_` between words, so this should be `credential_path`.

File: doc/man/task-sync.5.in [hunk 4]
```diff
@@ -85,6 +85,42 @@ Then configure Taskwarrior with:
...
+ Create your new role.
+
+ On the left sidebar, navigate to "Service accounts."
+
+ On the top menu bar within the "Service accounts" section, click "CREATE SERVICE ACCOUNT."
+ Provide an appropriate name and description for the new service account.
+ Select the role you just created and complete the service account creation process.
+
+ Now, in the Service Account dashboard, click into the new service account and select "keys" on the top menu bar.
+ Click on "ADD KEY" to create a new key or upload an existing key (a JSON key).
```
Comment (djmitche) [inline doc/man/task-sync.5.in]: How would uploading an existing key work here? Maybe "or upload an existing key" should be omitted here?

File: taskchampion/lib/src/server.rs [hunk 5]
```diff
@@ -170,6 +170,7 @@ pub unsafe extern "C" fn tc_server_new_sync(
 #[no_mangle]
 pub unsafe extern "C" fn tc_server_new_gcp(
     bucket: TCString,
+    credentialpath: Option<TCString>,
```
Comment (djmitche) [inline taskchampion/lib/src/server.rs]: Since this function is called from C++, it can't use Rust-specific types like `Option`. This will need to be just a plain `TCString`.

File: taskchampion/lib/src/server.rs [hunk 6]
```diff
@@ -180,6 +181,13 @@ pub unsafe extern "C" fn tc_server_new_gcp(
             //  - bucket ownership is transferred to this function
             let bucket = unsafe { TCString::val_from_arg(bucket) }.into_string()?;
 
+            // SAFETY:
+            //  - credentialpath is valid (promised by caller)
+            //  - credentialpath ownership is transferred to this function
+            let credentialpath = match credentialpath {
```
Comment (djmitche) [inline taskchampion/lib/src/server.rs]: Here I think you'll want to test whether `credentialpath == ""` and set the value passed to `ServerConfig::Gcp` to `None` for the empty string, and `Some(path)` otherwise.

File: taskchampion/taskchampion/src/server/cloud/gcp.rs [hunk 7]
```diff
@@ -25,9 +26,17 @@ fn is_http_error<T>(query: u16, res: &std::result::Result<T, http::Error>) -> bo
 }
 
 impl GcpService {
-    pub(in crate::server) fn new(bucket: String) -> Result<Self> {
+    pub(in crate::server) fn new(bucket: String, credentialpath:String) -> Result<Self> {
```
Comment (djmitche) [inline taskchampion/taskchampion/src/server/cloud/gcp.rs]: Since this is a pure Rust function, it should take the common Rust type, `Option<String>`, instead of special-casing the empty string. The conversion from the special-cased-empty-string to None is done in `tc_server_new_gcp`, which is a wrapper around `GcpService::new`.

File: taskchampion/taskchampion/src/server/cloud/gcp.rs [hunk 8]
```diff
@@ -242,12 +251,18 @@ mod tests {
     /// indicate anything else.
     fn make_service() -> Option<(GcpService, impl Fn(&str) -> Vec<u8>)> {
         let Ok(bucket) = std::env::var("GCP_TEST_BUCKET") else {
-            return None;
-        };
+            return None;};
+        
+        let Ok(credentialpath) = std::env::var("GCP_CREDENTIAL_PATH") else {
+            return None;};
+
```
Comment (djmitche) [inline taskchampion/taskchampion/src/server/cloud/gcp.rs]: Don't forget to run `cargo fmt` on your changes!

**Review by djmitche — CHANGES_REQUESTED** [review]
Looking good!

File: src/Context.cpp (hunk 3 above)
Comment (akashShanmugraj) [reply src/Context.cpp]: Okay!

File: doc/man/task-sync.5.in (hunk 4 above)
Comment (akashShanmugraj) [reply doc/man/task-sync.5.in]: In the Keys tab of every service account, there is an option to upload our own keys below the create new key option 

Screenshot 2024-01-22 at 6 48 45 AM

The user can upload his/her existing keys from the local storage

Screenshot 2024-01-22 at 6 48 51 AM

File: src/Context.cpp (hunk 3 above)
Comment (akashShanmugraj) [reply src/Context.cpp:286]: Okay! 
I have changed parameters throughout, just in case

File: taskchampion/lib/src/server.rs (hunk 5 above)
Comment (akashShanmugraj) [reply taskchampion/lib/src/server.rs]: Ah I forgot to push the new changes, what do you think of this?
```rust
#[no_mangle]
pub unsafe extern "C" fn tc_server_new_gcp( 
    bucket: TCString,
    credential_path_argument: TCString,
    encryption_secret: TCString,
    error_out: *mut TCString,
) -> *mut TCServer {
    wrap(
        || {
            // SAFETY:
            //  - bucket is valid (promised by caller)
            //  - bucket ownership is transferred to this function
            let bucket = unsafe { TCString::val_from_arg(bucket) }.into_string()?;

            // SAFETY:
            //  - credential_path is valid (promised by caller)
            //  - credential_path ownership is transferred to this function
            // TODOAK003: push changes after confirming with @djmitche
            let credential_path:String;

            if credential_path_argument.is_null() {
                credential_path = String::from("");
            } else {
                credential_path = unsafe { TCString::val_from_arg(credential_path_argument) }.into_string()?;
            }

            // SAFETY:  
 
```

File: taskchampion/taskchampion/src/server/cloud/gcp.rs (hunk 7 above)
Comment (akashShanmugraj) [reply taskchampion/taskchampion/src/server/cloud/gcp.rs]: So I need to change the datatype for both `bucket` and `credentialpath`?

File: taskchampion/taskchampion/src/server/cloud/gcp.rs (hunk 8 above)
Comment (akashShanmugraj) [reply taskchampion/taskchampion/src/server/cloud/gcp.rs]: Okay!

File: doc/man/task-sync.5.in (hunk 4 above)
Comment (djmitche) [reply doc/man/task-sync.5.in]: My question was, would someone who needs these instructions to know what to do be interested in uploading a cusotm X509 key? It seems unlikely, and I suspect anyone who did know how to generate an X509 key would not be reading these instructions.

File: taskchampion/lib/src/server.rs (hunk 5 above)
Comment (djmitche) [reply taskchampion/lib/src/server.rs]: Close, but `credential_path` should be an `Option<String>`.

File: taskchampion/taskchampion/src/server/cloud/gcp.rs (hunk 7 above)
Comment (djmitche) [reply taskchampion/taskchampion/src/server/cloud/gcp.rs]: No, Bucket can still be a String because it must be present.

File: taskchampion/lib/src/server.rs (hunk 6 above)
Comment (akashShanmugraj) [reply taskchampion/lib/src/server.rs]: I did something like this:

```rust
let credential_path:String;

if credential_path_argument.is_null() {
    credential_path = String::from("");
} else {
    credential_path = unsafe { TCString::val_from_arg(credential_path_argument) }.into_string()?;
}
```

And Gcp Serverconfig is initialised as it is already present, like
```rust
let server_config = ServerConfig::Gcp {
    bucket,
    credential_path: Some(credential_path),
    encryption_secret,
};
```

the reason I am passing `""` for the null case is because the credential_path in the authentication is checked against this empty string

```rust
let credentialpathstring = credential_path.clone().unwrap();
if credential_path.unwrap() == "" {
    config = rt.block_on(ClientConfig::default().with_auth())?;
} else {
    let credentials = rt.block_on(CredentialsFile::new_from_file(credentialpathstring))?;
    config = rt.block_on(ClientConfig::default().with_credentials(credentials))?;
}
```

File: taskchampion/lib/src/server.rs (hunk 5 above)
Comment (akashShanmugraj) [reply taskchampion/lib/src/server.rs]: Okay!

File: doc/man/task-sync.5.in (hunk 4 above)
Comment (akashShanmugraj) [reply doc/man/task-sync.5.in]: Yes, seems right. I'll change it to _Download/Create a new key_.
Plus, if someone is so experienced that they must upload a custom X509 key, then they must definitely know how to use it here!

File: taskchampion/taskchampion/src/server/cloud/gcp.rs (hunk 7 above)
Comment (akashShanmugraj) [reply taskchampion/taskchampion/src/server/cloud/gcp.rs]: okay 👍🏻

**Comment by akashShanmugraj** [comment]
@djmitche I guess the datatype changes here are correct?
[a5ee2e8](https://github.com/akashShanmugraj/taskwarrior/commit/a5ee2e8e9c1549db20cf2c1605112f8b5afd631f) and [227cfb2](https://github.com/akashShanmugraj/taskwarrior/commit/227cfb2d85f20328c0fe2285f6117de34ff7e616)

But I feel that these commits lack _Code Quality_ ?

File: taskchampion/taskchampion/src/server/config.rs [hunk 9]
```diff
@@ -27,6 +31,17 @@ pub enum ServerConfig {
         /// be any suitably un-guessable string of bytes.
         encryption_secret: Vec<u8>,
     },
+    /// A remote taskchampion-sync-server instance
+    #[cfg(feature = "server-gcp")]
+    Gcp {
+        /// Bucket in which to store the task data. This bucket must not be used for any other
+        /// purpose.
+        bucket: String,
+        credential_path: Option<String>,
```
Comment (djmitche) [inline taskchampion/taskchampion/src/server/config.rs:40]: Looks good -- please add a docstring!

File: taskchampion/lib/src/server.rs [hunk 10]
```diff
@@ -154,6 +154,69 @@ pub unsafe extern "C" fn tc_server_new_remote(
...
+/// for the description of the arguments.
+///
+/// On error, a string is written to the error_out parameter (if it is not NULL) and NULL is
+/// returned.  The caller must free this string.
+///
+/// The server must be freed after it is used - tc_replica_sync does not automatically free it.
+///
+/// ```c
+/// EXTERN_C struct TCServer *tc_server_new_gcp(struct TCString bucket,
+///                                       struct TCString credentia_path,
```
Comment (djmitche) [inline taskchampion/lib/src/server.rs]:
```suggestion
///                                       struct TCString credential_path,
```

File: taskchampion/lib/src/server.rs [hunk 11]
```diff
@@ -154,6 +154,69 @@ pub unsafe extern "C" fn tc_server_new_remote(
...
+            //  - credential_path ownership is transferred to this function
+
+            let credential_path: String;
+
+            if credential_path_argument.is_null() {
+                credential_path = String::from("");
+            } else {
+                credential_path =
+                    unsafe { TCString::val_from_arg(credential_path_argument) }.into_string()?;
+            }
```
Comment (djmitche) [inline taskchampion/lib/src/server.rs]: I talked about `is_null` earlier but I think this isn't actually what we want: TCStrings aren't actually allowed to be null. The `is_null` is a bit of a hack to handle some corner cases where there's a failure converting the TCString from one variant to the other. Specifically, there's no `tc_string_null` to create a null TCString from C.

Instead, I think we'll need to convert this to a `String`, and then check whether that is `""`.
```suggestion
            let credential_path =
                    unsafe { TCString::val_from_arg(credential_path_argument) }.into_string()?;
            let credential_path = if credential_path == "" {
                None
            } else {
                Some(credential_path)
            };
```

**Review by djmitche — COMMENTED** [review]
This is looking good!

The PR this is based on has now landed, so you can now make a PR against the https://github.com/GothenburgBitFactory/taskwarrior repository containing just your changes. That should be easier to review!

Also, please make sure to run `cargo fmt` and `cargo check` on the result.

File: taskchampion/lib/src/server.rs (hunk 11 above)
Comment (akashShanmugraj) [reply taskchampion/lib/src/server.rs]: Okay!

File: taskchampion/taskchampion/src/server/config.rs (hunk 9 above)
Comment (akashShanmugraj) [reply taskchampion/taskchampion/src/server/config.rs:40]: Okay!

File: taskchampion/lib/src/server.rs [hunk 12]
```diff
@@ -202,7 +202,7 @@ pub unsafe extern "C" fn tc_server_new_gcp(
                 .to_vec();
             let server_config = ServerConfig::Gcp {
                 bucket,
-                credential_path: Some(credential_path),
+                credential_path: credential_path.map(|cp| cp.to_string()),
```
Comment (akashShanmugraj) [inline taskchampion/lib/src/server.rs:205]: @djmitche I got a _Expected String, got Option<String>_ (probably because of None) so I changed it like this
Is this commit fine?

**Comment by akashShanmugraj** [comment]
Okay, I will open a new PR to the develop branch

File: taskchampion/lib/src/server.rs (hunk 12 above)
Comment (djmitche) [reply taskchampion/lib/src/server.rs:205]: It doesn't look right. What's the type of credential_path?

Comment (akashShanmugraj) [reply taskchampion/lib/src/server.rs:205]: `credential_path` is a String, `credential_path_argument` is a TCString

Comment (djmitche) [reply taskchampion/lib/src/server.rs:205]: Hm, looking up in this file I see 
```rust
let credential_path = if credential_path == "" {
                None
            } else {
                Some(credential_path)
            };
```
so that's a vlaue of type `Option<String>`. Don't forget that Rust allows new declarations of a variable name with a different type!

**Comment by djmitche** [comment]
Now at https://github.com/GothenburgBitFactory/taskwarrior/pull/3262
